### PR TITLE
Formatter supports escape characters

### LINKED ISSUE #35: Allow escaping of characters in format()
In French, hours are formatted like "8 h 40" and not "8:40". But in an arrow's format string, _h_ display directly the hours. Is it possible to add a key like _!_ to force _h_ to be displayed normally ("hh !h mm")?

### PR DESCRIPTION
**Pull request by remeika**
The `DateTimeFormatter` now supports brackets as escape characters. Based almost entirely off the work of @bsidhom.

```python
>>> formatter = arrow.formatter.DateTimeFormatter()
>>> dt = datetime(2018, 3, 9, 8, 40)

>>> fmt = "YYYY-MM-DD h [h] m"
>>> formatter.format(dt, fmt)
u'2018-03-09 8 h 40'

>>> fmt = "YYYY-MM-DD h [hello] m"
>>> formatter.format(dt, fmt)
u'2018-03-09 8 hello 40'

>>> fmt = "YYYY-MM-DD h [hello world] m"
>>> formatter.format(dt, fmt)
u'2018-03-09 8 hello world 40'
```

One note: I could not get coverage to pass on of my lines of code, and had to add a no cover pragma. If anyone has an idea why code coverage is unhappy with [`formatter.py:31`](https://github.com/crsmithdev/arrow/pull/688/files#diff-98102a005603dcf5eb33f3152fe5099eR30), I'd love to fix it.

Fixes #35

### CODE REVIEW

**Comment by systemcatch** [comment]
@remeika thanks for the PR, looks good from a quick glance over. Myself and Jad are pretty busy this week but hopefully can find some time to properly review it.

I wouldn't worry too much about the coverage problem since we're moving to pytest in #594 anyway.

File: arrow/formatter.py [hunk 1]
```diff
@@ -12,8 +12,9 @@
 class DateTimeFormatter(object):
 
     _FORMAT_RE = re.compile(
-        r"(YYY?Y?|MM?M?M?|Do|DD?D?D?|d?dd?d?|HH?|hh?|mm?|ss?|SS?S?S?S?S?|ZZ?Z?|a|A|X)"
+        r"(\[(?:(?=(?P<literal>\\\\\[|\\\\\]|[^]]))(?P=literal))*\]|YYY?Y?|MM?M?M?|Do|DD?D?D?|d?dd?d?|HH?|hh?|mm?|ss?|SS?S?S?S?S?|ZZ?Z?|a|A|X)"
```
Comment (systemcatch) [inline arrow/formatter.py]: Can you walk us through this new pattern?

File: arrow/formatter.py [hunk 2]
```diff
@@ -12,8 +12,9 @@
...
     )
+    _ESCAPED_BRACKET_RE = re.compile(r"\\\\(\[|\])")
```
Comment (systemcatch) [inline arrow/formatter.py]: Same for this one.

File: arrow/formatter.py [hunk 3]
```diff
@@ -25,6 +26,11 @@ def format(cls, dt, fmt):
 
     def _format_token(self, dt, token):
 
+        if token and token.startswith("[") and token.endswith("]"):
+            return self._ESCAPED_BRACKET_RE.sub(
+                lambda m: m.group(1), token[1:-1]
```
Comment (systemcatch) [inline arrow/formatter.py]: Could you explain exactly what the lambda function is doing here? Maybe adding a few comments to the source code would help.

File: arrow/formatter.py (hunk 1 above)
Comment (jadchaar) [reply arrow/formatter.py]: In particular, what is the purpose of the two pairs of escaped slashes?

**Comment by jadchaar** [comment]
Hey @remeika, any update on this? We'd love to merge this in, but we have a few clarifying questions.

**Comment by remeika** [comment]
@jadchaar Hey let me review the comments and reply later tonight!

**Comment by remeika** [comment]
Sorry for the late response here @jadchaar @systemcatch ! I've removed all handling of escaped brackets, which I think was motivated by [this comment](https://github.com/crsmithdev/arrow/issues/35#issuecomment-25513900) by @bsidhom; as I said, this basic implementation was his. 

But I don't think handling escaped brackets is really necessary, so this feature now does not handle them at all, which really simplifies things. I think this code is now ready to go!

**Comment by remeika** [comment]
@systemcatch As to why the last character of every escaped token is matching: I think this is an artifact of the workaround I'm using to emulate atomic capture groups. See the comment I added for details.

**Pull request merged**